File: db/questionnaire.py
#!/usr/bin/env python
'''
    medTurk (inspired by Amazon's Mechanical Turk) supports clinical research by using the 
    ingenuity of humans to convert unstructured clinical notes into structured data.

    Copyright (C) 2014 Innovation Center for Biomedical Informatics (ICBI)
                       Georgetown University <http://icbi.georgetown.edu/>
    
    Author: Robert M. Johnson "matt"
            <rmj49@georgetown.edu>
            <http://mattshomepage.com/>
            <@mattshomepage>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

from medturk.db import db
from bson import ObjectId







'''
    CREATE operations
'''
def create_questionnaire():
    return db.questionnaires.insert({'name'         : 'My New Questionnaire', 
                                     'description'  : 'TODO', 
                                     'questions'    : [], 
                                     'tags'         : []})

def create_uploaded_questionnaire(_questionnaire):
    db.questionnaires.insert(_questionnaire)

def create_question(_questionnaire_id):
    _id = ObjectId()
    question = {'_id' : _id, 'question' : '', 'type' : 'radio', 'frequency' : 'once', 'tag_ids' : [], 'choices' : [], 'triggers' : []}
    _questionnaire = get_questionnaire(_questionnaire_id)
    _questionnaire['questions'].append(question)
    db.questionnaires.save(_questionnaire)
    return question

def create_question_choice(_questionnaire_id, _question_id, _choice_name):

    _id = ObjectId()
    choice = {'_id' : _id, 'name' : _choice_name}

    _questionnaire = get_questionnaire(_questionnaire_id)
    _question_id = ObjectId(_question_id)
    for question in _questionnaire['questions']:
        if _question_id == question['_id']:
            question['choices'].append(choice)
            db.questionnaires.save(_questionnaire)
            break

    return choice


def create_question_trigger(_questionnaire_id, _question_id, _trigger_name, _trigger_case_sensitive):

    _id = ObjectId()
    trigger = {'_id' : _id, 'name' : _trigger_name, 'case_sensitive' : bool(_trigger_case_sensitive)}

    _questionnaire = get_questionnaire(_questionnaire_id)
    _question_id = ObjectId(_question_id)
    for question in _questionnaire['questions']:
        if _question_id == question['_id']:
            question['triggers'].append(trigger)
            db.questionnaires.save(_questionnaire)
            break

    return trigger

def create_question_tag(_questionnaire_id, _question_id, _tag_id):
    _questionnaire = get_questionnaire(_questionnaire_id)
    _question_id = ObjectId(_question_id)
    for question in _questionnaire['questions']:
        if _question_id == question['_id']:
            question['tag_ids'].append(_tag_id)
            db.questionnaires.save(_questionnaire)
            break

def create_questionnaire_tag(_questionnaire_id, _tag_name):
    _id = ObjectId()
    tag = {'_id' : _id, 'name' : _tag_name}
    _questionnaire = get_questionnaire(_questionnaire_id)
    _questionnaire['tags'].append(tag)
    db.questionnaires.save(_questionnaire)
    return tag












'''
    READ operations
'''
def get_questionnaires():
    return [m for m in db.questionnaires.find()]


def get_questionnaire(_questionnaire_id):
    return db.questionnaires.find_one({'_id' : ObjectId(_questionnaire_id)})











'''
    UPDATE operations
'''
def update_questionnaire_name(_questionnaire_id, _questionnaire_name):
    _questionnaire = get_questionnaire(_questionnaire_id)
    _questionnaire['name'] = _questionnaire_name
    db.questionnaires.save(_questionnaire)

def update_questionnaire_description(_questionnaire_id, _questionnaire_description):
    _questionnaire = get_questionnaire(_questionnaire_id)
    _questionnaire['description'] = _questionnaire_description
    db.questionnaires.save(_questionnaire)


def update_question_type(_questionnaire_id, _question_id, _question_type):
    _questionnaire = get_questionnaire(_questionnaire_id)
    _question_id = ObjectId(_question_id)
    for question in _questionnaire['questions']:
        if _question_id == question['_id']:
            question['type'] = _question_type
            db.questionnaires.save(_questionnaire)
            break


def update_question_frequency(_questionnaire_id, _question_id, _question_frequency):
    _questionnaire = get_questionnaire(_questionnaire_id)
    _question_id = ObjectId(_question_id)
    for question in _questionnaire['questions']:
        if _question_id == question['_id']:
            question['frequency'] = _question_frequency
            db.questionnaires.save(_questionnaire)
            break


def update_question_choice_name(_questionnaire_id, _question_id, _choice_id, _choice_name):
    _questionnaire = get_questionnaire(_questionnaire_id)

    # Convert to BSON
    _question_id = ObjectId(_question_id)
    _choice_id = ObjectId(_choice_id)

    for question in _questionnaire['questions']:
        if _question_id == question['_id']:
            for choice in question['choices']:
                if _choice_id == choice['_id']:
                    choice['name'] = _choice_name
                    db.questionnaires.save(_questionnaire)
                    break
            break



def update_question_text(_questionnaire_id, _question_id, _question_text):
    _questionnaire = get_questionnaire(_questionnaire_id)
    _question_id = ObjectId(_question_id)
    for question in _questionnaire['questions']:
        if _question_id == question['_id']:
            question['question'] = _question_text
            db.questionnaires.save(_questionnaire)
            break

def update_questionnaire_tag_name(_questionnaire_id, _tag_id, _tag_name):
    _questionnaire = get_questionnaire(_questionnaire_id)
    _tag_id = ObjectId(_tag_id)
    for tag in _questionnaire['tags']:
        if _tag_id == tag['_id']:
            tag['name'] = _tag_name
            db.questionnaires.save(_questionnaire)
            break









'''
    DELETE operations
'''
def delete_questionnaire_tag(_questionnaire_id, _tag_id):
    
    _questionnaire = get_questionnaire(_questionnaire_id)

    index = -1
    _tag_id = ObjectId(_tag_id)

    for i,tag in enumerate(_questionnaire['tags']):
        if tag['_id'] == _tag_id:
            index = i
            break
    
    if index > -1:
        del _questionnaire['tags'][index]
        db.questionnaires.save(_questionnaire)


def delete_question_tag(_questionnaire_id, _question_id, _tag_id):
    
    _questionnaire = get_questionnaire(_questionnaire_id)

    index = -1
    _question_id = ObjectId(_question_id)

    for i,question in enumerate(_questionnaire['questions']):
        if question['_id'] == _question_id:
            for j,tag_id in enumerate(_questionnaire['questions'][i]['tag_ids']):
                if tag_id == _tag_id:
                    del _questionnaire['questions'][i]['tag_ids'][j]
                    db.questionnaires.save(_questionnaire)
                    return


def delete_question_trigger(_questionnaire_id, _question_id, _trigger_id):
    
    _questionnaire = get_questionnaire(_questionnaire_id)

    index = -1
    _question_id  = ObjectId(_question_id)
    _trigger_id   = ObjectId(_trigger_id)

    for i,question in enumerate(_questionnaire['questions']):
        if question['_id'] == _question_id:
            for j,trigger in enumerate(_questionnaire['questions'][i]['triggers']):
                if trigger['_id'] == _trigger_id:
                    del _questionnaire['questions'][i]['triggers'][j]
                    db.questionnaires.save(_questionnaire)
                    return

def delete_question_choice(_questionnaire_id, question_id, choice_id):
    
    _questionnaire = get_questionnaire(_questionnaire_id)

    index = -1
    question_id = ObjectId(question_id)
    choice_id   = ObjectId(choice_id)

    for i,question in enumerate(_questionnaire['questions']):
        if question['_id'] == question_id:
            for j,choice in enumerate(_questionnaire['questions'][i]['choices']):
                if choice['_id'] == choice_id:
                    del _questionnaire['questions'][i]['choices'][j]
                    db.questionnaires.save(_questionnaire)
                    return

def delete_question(_questionnaire_id, _question_id):
    
    _questionnaire = get_questionnaire(_questionnaire_id)

    index = -1
    _question_id = ObjectId(_question_id)

    for i,question in enumerate(_questionnaire['questions']):
        if question['_id'] == _question_id:
            index = i
            break
    
    if index > -1:
        del _questionnaire['questions'][index]
        db.questionnaires.save(_questionnaire)



def delete_questionnaire(_questionnaire_id):
    db.questionnaires.remove({'_id' : ObjectId(_questionnaire_id)})

File: node_js/db/questionnaire.js
var mongoskin  = require('mongoskin')
var db = require('./database.js').db()

module.exports = {

	generate_id: function() {
		return new mongoskin.ObjectID()
	},

	update_question_type: function(_questionnaire_id, _question_id, _type, _callback, _error_callback, _passthrough) {

		_questionnaire_id = new mongoskin.ObjectID(_questionnaire_id)
		_question_id      = new mongoskin.ObjectID(_question_id)

		var arg1 = {'_id' : _questionnaire_id, "questions" : {$elemMatch : {'_id' : _question_id}}}
		var arg2 = { $set : {"questions.$.type" : _type}}

		db.collection('questionnaires').update(arg1, arg2, function(_err, _result) {
			if (_err) {
				_error_callback(_err, _passthrough)
			}
			else if (_result) {
				_callback(_passthrough)
			}
			else {
				_error_callback(_err, _passthrough)
			}
		})
	},


	update_question_frequency: function(_questionnaire_id, _question_id, _frequency, _callback, _error_callback, _passthrough) {

		_questionnaire_id = new mongoskin.ObjectID(_questionnaire_id)
		_question_id      = new mongoskin.ObjectID(_question_id)

		var arg1 = {'_id' : _questionnaire_id, "questions" : {$elemMatch : {'_id' : _question_id}}}
		var arg2 = { $set : {"questions.$.frequency" : _frequency}}

		db.collection('questionnaires').update(arg1, arg2, function(_err, _result) {
			if (_err) {
				_error_callback(_err, _passthrough)
			}
			else if (_result) {
				_callback(_passthrough)
			}
			else {
				_error_callback(_err, _passthrough)
			}
		})
	},


	update_question_choice_name: function(_questionnaire_id, _question_id, _choice_id, _name, _callback, _error_callback, _passthrough) {

		_questionnaire_id = new mongoskin.ObjectID(_questionnaire_id)
		_question_id      = new mongoskin.ObjectID(_question_id)

		// [1] Only retrieve the question portion of the questionnaire
		// [2] Update choice porition of the question
		// [3] Update the question subdocument in the questionnaire

		// Query for the questionnaire
		var arg1 = {'_id' : _questionnaire_id}

		// But only return the question
		var arg2 = {_id: 0, questions: {$elemMatch: {_id:  _question_id}}}

		// Fetch the question
		db.collection('questionnaires').findOne(arg1, arg2, function(_err, _result) {
			if (_err) {
				_error_callback(_err, _passthrough)
			}
			else if (_result) {
				
				var question = _result.questions[0]
				var choice = undefined

				for (i = 0; i < question.choices.length; i++) {
					if (question.choices[i]._id == _choice_id) {
						choice = question.choices[i]
						break
					}
				}

				if (choice == undefined) {
					_error_callback(null, _passthrough)
				}
				else {

					// Now we can update the choice name
					choice.name = _name

					arg1 = {'_id' : _id, "questions" : {$elemMatch : {'_id' : question_id}}}
					arg2 = { $set : {"questions.$" : question}}

					db.collection('questionnaires').update(arg1, arg2, function(_err, _result) {
						if (_err) {
							_error_callback(_err, _passthrough)
						}
						else if (_result) {
							_callback(_passthrough)
						}
						else {
							_error_callback(_err, _passthrough)
						}
						
					})
				}
				
			}
			else {
				_error_callback(null, _passthrough)
			}
		})
	},


	update_question_name: function(_questionnaire_id, _question_id, _name, _callback, _error_callback, _passthrough) {
		
		_questionnaire_id = new mongoskin.ObjectID(_questionnaire_id)
		_question_id      = new mongoskin.ObjectID(_question_id)

		var arg1 = {'_id' : _questionnaire_id, "questions" : {$elemMatch : {'_id' : _question_id}}}
		var arg2 = { $set : {"questions.$.question" : _name}}

		db.collection('questionnaires').update(arg1, arg2, function(_err, _result) {
			if (_err) {
				_error_callback(_err, _passthrough)
			}
			else if (_result) {
				_callback(_passthrough)
			}
			else {
				_error_callback(null, _passthrough)
			}
		})
	},


	update_tag_name: function(_questionnaire_id, _tag_id, _name, _callback, _error_callback, _passthrough) {

		_tag_id = new mongoskin.ObjectID(_tag_id)

		var arg1 = {'_id' : new mongoskin.ObjectID(_questionnaire_id), "tags" : {$elemMatch : {'_id' : _tag_id}}}
		var arg2 = { $set : {"tags.$.name" : _name}}

		db.collection('questionnaires').update(arg1, arg2, function(_err, _result) {
			if (_err) {
				_error_callback(_err, _passthrough)
			}
			else if (_result) {
				_callback(_passthrough)
			}
			else {
				_error_callback(null, _passthrough)
			}
		})
	},


	update_name: function(_questionnaire_id, _name, _callback, _error_callback, _passthrough) {

		var arg1 = {'_id' : new mongoskin.ObjectID(_questionnaire_id)}
		var arg2 = {'$set' : {'name' : _name}}

		db.collection('questionnaires').update(arg1, arg2, function(_err, _result) {
			if (_err) {
				_error_callback(_err, _passthrough)
			}
			else if (_result) {
				_callback(_passthrough)
			}
			else {
				_error_callback(null, _passthrough)
			}
			
		})
	},


	update_description: function(_questionnaire_id, _description, _callback, _error_callback, _passthrough) {

		var arg1 = {'_id' : new mongoskin.ObjectID(_questionnaire_id)}
		var arg2 = {'$set' : {'description' : _description}}

		db.collection('questionnaires').update(arg1, arg2, function(_err, _result) {
			if (_err) {
				_error_callback(_err, _passthrough)
			}
			else if (_result) {
				_callback(_passthrough)
			}
			else {
				_error_callback(null, _passthrough)
			}
			
		})
	},


	insert_questionnaire: function(_questionnaire, _callback, _error_callback, _passthrough) {

		db.collection('questionnaires').insert(_questionnaire, function(_err, _result) {
			if (_err) {
				_error_callback(_err, _passthrough)
			}
			else {
				_callback({'_id' : _result[0]._id}, _passthrough)
			}
		})
	},

	get_by_id: function(_questionnaire_id, _callback, _error_callback, _passthrough) {

		var _query = {'_id' : new mongoskin.ObjectID(_questionnaire_id)}

		db.collection('questionnaires').findOne(_query, function(_err, _result) {
			if (_err) {
				_error_callback(_err, _passthrough)
			}
			else {
				_callback(_result, _passthrough)
			}
		})
	},


	get_all_questionnaires: function(_callback, _error_callback, _passthrough) {
		db.collection('questionnaires').find().toArray(function(_err, _result) {
			if (_err) {
				_error_callback(_err, _passthrough)
			}
			else {
				_callback(_result, _passthrough)
			}
		})
	},


	create_tag: function(_questionnaire_id, _tag_name, _callback, _error_callback, _passthrough) {

		var arg1 = {'_id' : new mongoskin.ObjectID(_questionnaire_id)}

		db.collection('questionnaires').findOne(arg1, function(err, result) {
			if (err) throw err
			
			// Add this tag to the current collection
			var tag = {'_id' : new mongoskin.ObjectID(), 'name' : _tag_name}
			result.tags.push(tag)

			var arg2 = {'$set' : {'tags' : result.tags}}

			db.collection('questionnaires').update(arg1, arg2, function(_err, _result) {
				if (_err) {
				_error_callback(_err, _passthrough)
			}
			else if (_result) {
				_callback(tag, _passthrough)
			}	
			else {
				_error_callback(null, _passthrough)
			}
			})
		})
	},


	create_trigger: function(_questionnaire_id, _question_id, _trigger_name, _case_sensitive, _callback, _error_callback, _passthrough) {
		
		_questionnaire_id = new mongoskin.ObjectID(_questionnaire_id)
		_question_id      = new mongoskin.ObjectID(_question_id)

		var doc  = {'_id' : new mongoskin.ObjectID(), 'name' : _trigger_name, 'case_sensitive' : _case_sensitive}

		var arg1 = {'_id' : _questionnaire_id, "questions" : {$elemMatch : {'_id' : _question_id}}}
		var arg2 = { $push: { 'questions.$.triggers' : doc } }

		db.collection('questionnaires').update(arg1, arg2, function(_err, _result) {
			if (_err) {
				_error_callback(_err, _passthrough)
			}
			else if (_result) {
				_callback(doc, _passthrough)
			}	
			else {
				_error_callback(null, _passthrough)
			}
		})
	},


	insert_question_tag_id: function(_questionnaire_id, _question_id, _tag_id, _callback, _error_callback, _passthrough) {
		
		_questionnaire_id = new mongoskin.ObjectID(_questionnaire_id)
		_question_id      = new mongoskin.ObjectID(_question_id)
		_tag_id           = new mongoskin.ObjectID(_tag_id)

		var doc  = _tag_id

		var arg1 = {'_id' : _questionnaire_id, "questions" : {$elemMatch : {'_id' : _question_id}}}
		var arg2 = { $push: { 'questions.$.tag_ids' : doc } }

		db.collection('questionnaires').update(arg1, arg2, function(_err, _result) {
			if (_err) {
				_error_callback(_err, _passthrough)
			}
			else if (_result) {
				_callback(doc, _passthrough)
			}	
			else {
				_error_callback(null, _passthrough)
			}
		})
	},

	create_question_choice: function(_questionnaire_id, _question_id, _name, _callback, _error_callback, _passthrough) {

		_questionnaire_id = new mongoskin.ObjectID(_questionnaire_id)
		_question_id      = new mongoskin.ObjectID(_question_id)

		var doc  = {'_id' : new mongoskin.ObjectID(), 'name' : _name}

		var arg1 = {'_id' : _questionnaire_id, "questions" : {$elemMatch : {'_id' : _question_id}}}
		var arg2 = { $push: { 'questions.$.choices' : doc } }

		db.collection('questionnaires').update(arg1, arg2, function(_err, _result) {
			if (_err) {
				_error_callback(_err, _passthrough)
			}
			else if (_result) {
				_callback(doc, _passthrough)
			}	
			else {
				_error_callback(null, _passthrough)
			}
		})
	},


	create_question: function(_questionnaire_id, _question, _type, _frequency, _tag_ids, _choices, _triggers, _callback, _error_callback, _passthrough) {

		var doc = {'_id'       : new mongoskin.ObjectID(), 
				  'question'   : _question, 
				  'type'       : _type, 
				  'frequency'  : _frequency, 
				  'tag_ids'    : _tag_ids, 
				  'choices'    : _choices, 
				  'triggers'   : _triggers}
		

		var arg1 = {'_id' : new mongoskin.ObjectID(_questionnaire_id)}
		var arg2 = { $push: { 'questions' : doc } }

		db.collection('questionnaires').update(arg1, arg2, function(_err, _result) {
			if (_err) {
				_error_callback(_err, _passthrough)
			}
			else if (_result) {
				_callback(doc, _passthrough)
			}	
			else {
				_error_callback(null, _passthrough)
			}
		})
	},


	create_questionnaire: function(_questionnaire_name, _questionnaire_description, _questions, _tags, _callback, _error_callback, _passthrough) {
		
		var doc = {'name' : _questionnaire_name, 
				  'description' : _questionnaire_description, 
				  'questions' : _questions, 
				  'tags' :_tags}

		db.collection('questionnaires').insert(doc, function(_err, _result) {
			if (_err) {
				_error_callback(_err, _passthrough)
			}
			else if(_result) {
				_callback(_result[0], _passthrough)
			}
			else {
				_error_callback(null, _passthrough)
			}
		})
	},



	delete_question: function(_questionnaire_id, _question_id, _callback, _error_callback, _passthrough) {
		
		var arg1 = {'_id' : new mongoskin.ObjectID(_questionnaire_id)}
		var arg2 = {$pull : {'questions' : {'_id': new mongoskin.ObjectID(_question_id)} }}

		db.collection('questionnaires').update(arg1, arg2, function(_err, _result) {
			if (_err) {
				_error_callback(_err, _passthrough)
			}
			else if(_result) {
				_callback(_passthrough)
			}
			else {
				_error_callback(null, _passthrough)
			}
		})
	},


	delete_tag: function(_questionnaire_id, _tag_id, _callback, _error_callback, _passthrough) {
		
		var arg1 = {'_id' : new mongoskin.ObjectID(_questionnaire_id)}
		var arg2 = {$pull : {'tags' : {'_id': new mongoskin.ObjectID(_tag_id)} }}

		db.collection('questionnaires').update(arg1, arg2, function(_err, _result) {
			if (_err) {
				_error_callback(_err, _passthrough)
			}
			else if(_result) {
				_callback(_passthrough)
			}
			else {
				_error_callback(null, _passthrough)
			}
		})
	},


	delete_questionnaire: function(_questionnaire_id, _callback, _error_callback, _passthrough) {

		var arg = {'_id' : new mongoskin.ObjectID(_questionnaire_id)}

		db.collection('questionnaires').remove(arg, function(_err, _result) {
			if (_err) {
				_error_callback(_err, _passthrough)
			}
			else if(_result) {
				_callback(_passthrough)
			}
			else {
				_error_callback(null, _passthrough)
			}
		})
	},


	delete_question_choice: function(_questionnaire_id, _question_id, _choice_id, _callback, _error_callback, _passthrough) {

		_questionnaire_id = new mongoskin.ObjectID(_questionnaire_id)
		_question_id      = new mongoskin.ObjectID(_question_id)

		// [1] Only retrieve the question portion of the questionnaire
		// [2] Update choice porition of the question
		// [3] Update the question subdocument in the questionnaire

		// Query for the questionnaire
		var arg1 = {'_id' : _questionnaire_id}

		// But only return the question
		var arg2 = {_id: 0, questions: {$elemMatch: {_id:  _question_id}}}

		// Fetch the question
		db.collection('questionnaires').findOne(arg1, arg2, function(_err, _result) {
			if (_err) {
				_error_callback(_err, _passthrough)
			}
			else if (_result) {
				var question = _result.questions[0]
				var choice_index = undefined

				for (i = 0; i < question.choices.length; i++) {
					if (question.choices[i]._id == _choice_id) {
						choice_index = i
			
						break
					}
				}


				if (choice_index == undefined) {
					_error_callback(null, _passthrough)
				}
				else {

					// Now we can remove the choice
					question.choices.splice(choice_index, 1)

					arg1 = {'_id' : _questionnaire_id, "questions" : {$elemMatch : {'_id' : _question_id}}}
					arg2 = { $set : {"questions.$" : question}}

					db.collection('questionnaires').update(arg1, arg2, function(_err, _result) {
						if (_err) {
							_error_callback(_err, _passthrough)
						}
						else if (_result) {
							_callback(_passthrough)
						}
						else {
							_error_callback(null, _passthrough)
						}
						
					})
				}
				
			}
			else {
				// Couldn't find it
				_error_callback(null, _passthrough)
			}
		})

	},


	delete_question_trigger: function(_questionnaire_id, _question_id, _trigger_id, _callback, _error_callback, _passthrough) {

		_questionnaire_id = new mongoskin.ObjectID(_questionnaire_id)
		_question_id 	  = new mongoskin.ObjectID(_question_id)

		// [1] Only retrieve the question portion of the questionnaire
		// [2] Update choice porition of the question
		// [3] Update the question subdocument in the questionnaire

		// Query for the questionnaire
		var arg1 = {'_id' : _questionnaire_id}

		// But only return the question
		var arg2 = {_id: 0, questions: {$elemMatch: {_id:  _question_id}}}

		// Fetch the question
		db.collection('questionnaires').findOne(arg1, arg2, function(_err, _result) {
			if (_err) {
				_error_callback(_err, _passthrough)
			}
			else if (_result) {
				var question = _result.questions[0]
				var trigger_index = undefined

				for (i = 0; i < question.triggers.length; i++) {
					if (question.triggers[i]._id == _trigger_id) {
						trigger_index = i
						break
					}
				}


				if (trigger_index == undefined) {
					_error_callback(null, _passthrough)
				}
				else {

					// Now we can remove the choice
					question.triggers.splice(trigger_index, 1)

					arg1 = {'_id' : _questionnaire_id, "questions" : {$elemMatch : {'_id' : _question_id}}}
					arg2 = { $set : {"questions.$" : question}}

					db.collection('questionnaires').update(arg1, arg2, function(err, result) {
						if (_err) {
							_error_callback(_err, _passthrough)
						}
						else if (result) {
							_callback(_passthrough)
						}
						else {
							_error_callback(null, _passthrough)
						}
						
					})
				}
				
			}
			else {
				// Couldn't find it
				_error_callback(null, _passthrough)
			}
		})
	},


	delete_question_tag: function(_questionnaire_id, _question_id, _tag_id, _callback, _error_callback, _passthrough) {

		_questionnaire_id = new mongoskin.ObjectID(_questionnaire_id)
		_question_id      = new mongoskin.ObjectID(_question_id)

		// [1] Only retrieve the question portion of the questionnaire
		// [2] Update tag portion of the question
		// [3] Update the question subdocument in the questionnaire

		// Query for the questionnaire
		var arg1 = {'_id' : _questionnaire_id}

		// But only return the question
		var arg2 = {_id: 0, questions: {$elemMatch: {_id:  _question_id}}}

		// Fetch the question
		db.collection('questionnaires').findOne(arg1, arg2, function(_err, _result) {
			if (_err) {
				_error_callback(null, _passthrough)
			}
			else if (_result) {
				// RMJ
				var question = _result.questions[0]
				var tag_index = undefined

				for (i = 0; i < question.tag_ids.length; i++) {
					if (question.tag_ids[i] == _tag_id) {
						tag_index = i
						break
					}
				}


				if (tag_index == undefined) {
					return res.sendStatus(404)
				}
				else {

					// Now we can remove the tag
					question.tag_ids.splice(tag_index, 1)

					arg1 = {'_id' : _questionnaire_id, "questions" : {$elemMatch : {'_id' : _question_id}}}
					arg2 = { $set : {"questions.$" : question}}

					db.collection('questionnaires').update(arg1, arg2, function(_err, _result) {
						if (_err) {
							_error_callback(_err, _passthrough)
						}
						else if (_result) {
							_callback(_passthrough)
						}
						else {
							_error_callback(null, _passthrough)
						}
						
					})
				}
				
			}
			else {
				// Couldn't find it
				_error_callback(null, _passthrough)
			}
		})
	}


}